sentry-android-ndk/src/main/jni/sentry.c

```diff
@@ -11,6 +11,190 @@ struct transport_options {
+    const char *charKey = (*env)->GetStringUTFChars(env, key, 0);
+
+    sentry_remove_extra(charKey);
+}
+
+JNIEXPORT void JNICALL Java_io_sentry_android_ndk_NdkScopeObserver_nativeSetUser(
+        JNIEnv *env,
+        jclass cls,
+        jstring id,
+        jstring email,
+        jstring ipAddress,
+        jstring username) {
+    sentry_value_t user = sentry_value_new_object();
+    if (id) {
+        const char *charId = (*env)->GetStringUTFChars(env, id, 0);
+        sentry_value_set_by_key(user, "id", sentry_value_new_string(charId));
+    }
+    if (email) {
+        const char *charEmail = (*env)->GetStringUTFChars(env, email, 0);
+        sentry_value_set_by_key(
+                user, "email", sentry_value_new_string(charEmail));
+    }
+    if (ipAddress) {
+        const char *charIpAddress = (*env)->GetStringUTFChars(env, ipAddress, 0);
+        sentry_value_set_by_key(
+                user, "ip_address", sentry_value_new_string(charIpAddress));
+    }
+    if (username) {
+        const char *charUsername = (*env)->GetStringUTFChars(env, username, 0);
+        sentry_value_set_by_key(
+                user, "username", sentry_value_new_string(charUsername));
+    }
+    sentry_set_user(user);
+}
+
+JNIEXPORT void JNICALL Java_io_sentry_android_ndk_NdkScopeObserver_nativeRemoveUser(
+        JNIEnv *env,
+        jclass cls) {
+    sentry_remove_user();
+}
+
+JNIEXPORT void JNICALL Java_io_sentry_android_ndk_NdkScopeObserver_nativeAddBreadcrumb(
+        JNIEnv *env,
+        jclass cls,
+        jstring level,
+        jstring message,
+        jstring category,
+        jstring type,
+        jstring timestamp,
+        jobject data) {
+    if (!level && !message && !category && !type) {
+        return;
+    }
+    const char *charMessage = NULL;
+    if (message) {
+        charMessage = (*env)->GetStringUTFChars(env, message, 0);
+    }
+    const char *charType = NULL;
+    if (type) {
+        charType = (*env)->GetStringUTFChars(env, type, 0);
+    }
+    sentry_value_t crumb
+            = sentry_value_new_breadcrumb(charType, charMessage);
+
+    if (category) {
+        const char *charCategory = (*env)->GetStringUTFChars(env, category, 0);
+        sentry_value_set_by_key(
+                crumb, "category", sentry_value_new_string(charCategory));
+    }
+    if (level) {
+        const char *charLevel = (*env)->GetStringUTFChars(env, level, 0);
+        sentry_value_set_by_key(
+                crumb, "level", sentry_value_new_string(charLevel));
+    }
+
+    // overwrite timestamp that is already created on sentry_value_new_breadcrumb
+    if (timestamp) {
+        const char *charTimestamp = (*env)->GetStringUTFChars(env, timestamp, 0);
+        sentry_value_set_by_key(
+                crumb, "timestamp", sentry_value_new_string(charTimestamp));
+    }
+
+    // TODO: make this a method
+    if (data) {
```

Comment:
@Swatinem do you know an easier way to do this? from Map<String, Object> on java to sentry-native?

Reply:
damn, I have no idea. But if you put this into a method, you can at least `goto fail` and have a few finalizers (those `DeleteLocalRef`) there. That should at least take care of the nesting.

Those `jclass`/`jmethodID` things can probably be done someplace else and cached inside a struct. I hope the pointers returned from there are stable across GC?

I can imagine something like:
```
struct {
jclass map;
jmethodID map_entrySet;
jclass set;
jmethodID set_iterator;
…
}
```

Also, why do you need `entrySet`, can’t you get an iterator from `Map` directly?

Reply:
for the last question, no, https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--

the suggestions above really make sense, I will experiment with just parsing the map as a String, let's see how it goes, last case I'd do this way, thanks